gossip/minimal/goshawk.go

```diff
@@ -56,6 +61,132 @@ type FetchOptions struct {
 	BatchSize int
 	// Number of concurrent fetcher workers to run.
 	ParallelFetch int
+	// Manage hub retrieval state persistence.
+	State         ScanStateManager
+	FlushInterval time.Duration
+}
+
+// ScanStateManager controls hub scanning state, with the intention of allowing
+// scanning to resume where it was left off across restarts etc.
+type ScanStateManager interface {
+	// GetHubIndex returns the index at which scanning should commence for the given hub URL.
+	GetHubIndex(hubURL string) int64
+	// UpdateHubIndex indicates that future scanning of the given hub should start from at least nextIndex.
+	UpdateHubIndex(hubURL string, nextIndex int64)
+	// Flush ensures state is stored.
+	Flush(ctx context.Context) error
+}
+
+// ScanState holds a (mutex-protected) map of hub URL to next index to scan from.
+type ScanState struct {
+	Mu   sync.Mutex
+	Next map[string]int64
+}
+
+// GetHubIndex returns the index at which scanning should commence for the given hub URL.
+func (s *ScanState) GetHubIndex(hubURL string) int64 {
+	s.Mu.Lock()
+	defer s.Mu.Unlock()
+	return s.Next[hubURL]
+}
+
+// UpdateHubIndex indicates that future scanning of the given hub should start from at least nextIndex.
+func (s *ScanState) UpdateHubIndex(hubURL string, nextIndex int64) {
+	s.Mu.Lock()
+	defer s.Mu.Unlock()
+	cur := s.Next[hubURL]
+	if nextIndex > cur {
+		s.Next[hubURL] = nextIndex
+	}
+}
+
+// memoryStateManager is an implementation of the ScanStateManager interface
+// that does not persist any data; it is intended for use as a fallback and in
+// testing.
+type memoryStateManager struct {
+	ScanState
+}
+
+// Flush does nothing for a memoryStateManager.
+func (m *memoryStateManager) Flush(ctx context.Context) error {
+	// No persistent storage.
+	return nil
+}
+
+// FileStateManager is an implementation of the ScanStateManager interface
+// that stores the scan state in a writable flat file.
+type FileStateManager struct {
+	ScanState
+	filename string
+}
+
+// NewFileStateManager creates a FileStateManager from the contents of the given
+// filename, which can be be empty but should be over-writable.
+func NewFileStateManager(filename string) (*FileStateManager, error) {
+	f := FileStateManager{
+		ScanState: ScanState{Next: make(map[string]int64)},
+		filename:  filename,
+	}
+
+	file, err := os.Open(filename)
```

Comment:
I think the read logic should be broken out similarly to flush(). Not necessarily in the public API but it might help with testing if it was.

Reply:
Done.